### PR TITLE
core/scheduler: verify duty pubkeys in resolveDuties

### PR DESCRIPTION
**Pull request by gsora**
Do not accept duties for which validator keys are not what we expect.

category: feature
ticket: #1922

Replaces #2194.

Closes #2194.

### CODE REVIEW

File: testutil/beaconmock/options.go [hunk 1]
```diff
@@ -242,6 +242,58 @@ func WithSlotsPerEpoch(slotsPerEpoch int) Option {
 	}
 }
 
+// WithDutiesPubkeys configures the mock to return attestation, sync committee and proposer
+// duties with the prpovided public key.
+func WithDutiesPubkeys(key eth2p0.BLSPubKey) Option {
```
Comment (corverroos) [inline testutil/beaconmock/options.go:247]: adding this to beaconmock for one specific test feels a bit much, rather only do this in the test itself.

File: core/scheduler/scheduler_test.go [hunk 2]
```diff
@@ -369,6 +369,71 @@ func TestScheduler_GetDuty(t *testing.T) {
 	require.NoError(t, sched.Run())
 }
 
+func TestScheduler_GetDutyDifferentPubkeys(t *testing.T) {
```
[the remark this answers is not in the document]
Comment (gsora) [reply core/scheduler/scheduler_test.go]: Implemented as discussed!

File: core/scheduler/scheduler_internal_test.go [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+		beaconmock.WithValidatorSet(valSet),
+		beaconmock.WithGenesisTime(t0),
+		beaconmock.WithDeterministicAttesterDuties(0),
+		beaconmock.WithDeterministicProposerDuties(0),
+		beaconmock.WithDeterministicSyncCommDuties(2, 2),
+		beaconmock.WithSlotsPerEpoch(1),
+	)
+	require.NoError(t, err)
+	_, err = eth2Cl.SlotDuration(ctx)
+	require.NoError(t, err)
```
Comment (corverroos) [inline core/scheduler/scheduler_internal_test.go:39]: what is the purpose of this?

File: core/scheduler/scheduler_internal_test.go [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+	oldProposerFunc := eth2Cl.ProposerDutiesFunc
+
+	eth2Cl.AttesterDutiesFunc = func(ctx context.Context, epoch eth2p0.Epoch, indices []eth2p0.ValidatorIndex) ([]*eth2v1.AttesterDuty, error) {
+		res, err := oldAttesterFunc(ctx, epoch, indices)
+		if err != nil {
+			return nil, err
+		}
+
+		for idx := 0; idx < len(res); idx++ {
+			elem := res[idx]
```
Comment (corverroos) [inline core/scheduler/scheduler_internal_test.go:54]: nit: slightly more readable: `res[idx].Pubkey = testutil.RandomEth2PubKey(t)`

File: core/scheduler/scheduler_internal_test.go [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+	require.NoError(t, err)
+	_, err = eth2Cl.SlotDuration(ctx)
+	require.NoError(t, err)
+
+	key := testutil.RandomEth2PubKey(t)
+
+	oldAttesterFunc := eth2Cl.AttesterDutiesFunc
+	oldSyncFunc := eth2Cl.SyncCommitteeDutiesFunc
+	oldProposerFunc := eth2Cl.ProposerDutiesFunc
+
```
Comment (corverroos) [inline core/scheduler/scheduler_internal_test.go:46]: add comment here: `Randomize duty pubkeys`

File: core/scheduler/scheduler_internal_test.go [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+	}
+
+	return sched, schedVals
+}
+
+func TestResolveAttDuties(t *testing.T) {
+	sched, schedVals := setupScheduler(t)
+
+	require.ErrorContains(t, sched.resolveAttDuties(context.Background(), core.Slot{
+		Slot:          1,
```
Comment (corverroos) [inline core/scheduler/scheduler_internal_test.go:115]: can probably omit some/most `core.Slot` fields here to make it less verbose

**Review by corverroos — APPROVED** [review]

**Comment by corverroos** [comment]
LGTM!

**Pull request merged**